### ISSUE TITLE
RF: metadata layout, storage, aggregation, and access procedure

### ISSUE THREAD

**Issue by mih** [datalad/datalad#1287]
This will eventually turn into a PR hat implements the described changes.

## Major aims

- change the previous concept so metadata aggregation has minimal impact on the Git history of a dataset, in particular no metadata should be directly committed in a superdataset in order to limit the disaster scope in case of "sensible" metadata leakage. This is achieved by switching to annex'ed files as metadata store (for basic dataset metadata this is not the case, see below)
- minimize impact on superdataset performance when subdatasets have huge metadata sets. we need to find a good tradeoff between number of required inodes and granularity of storage (ideally the pieces are relatively stable over time for typical datasets)
- mimimize, and ideally entirely avoid, the need to anyhow alter metadata files from sub(sub(sub))datasets when aggregating upwards. At the same time enable filtering options for metadata aggregation in superdatasets, in particular avoiding the need for heuristics and needless metadata tagging to enable filtering when everything is stored in a single file
- ~~avoid fileformat debate/selection/support for basic dataset metadata by using git annex's own metadata API. This will maintain the concept of *immediate metadata availability* after having obtained no more than the Git repo of a dataset~~ PR #1630 provides a hybrid approach to this. It allows manipulation via a uniform API (gitannex and dataset metadata), but keeps storage of dataset-level metadata in a file in Git

## Storage concept

- no conflict with previous scheme, old metadata can be kept for as long as desired by dataset author
- two locations are used
    - `.datalad/metadata/info`: JSON file providing basic properties for all aggregated metadata files in the store (see next). This file is always present, even when no subdatasets exist.
        - **Basic dataset metadata is attached to this file using `git annex metadata`**. We might need to include some info on subdatasets in here too, in order to have useful minimal metadata for subdatasets after "`git clone`"
        - Basic dataset metadata of **subdataset** could be aggregated into the **content** of this file in a superdataset
        - this is an annex'ed file: this is a requirement of using git annex metadata, but also completes the detachment of aggregated metadata content from a superdataset.
    - `.datalad/meta/objects`: directory of annex'ed files for all aggregated metadata from subdatasets.
        - filenames could be something like: `<hash>-<categorylabel>`, where `<hash>` is the SHA of the repository the metadata was produced from (or maybe something shorter than the SHA, but with equally low collision probability in this context), and `<categorylabel>` is an identifier for the "type" of metadata -- in order to distinguish multiple metadata sets produced from the same dataset state
        - the content of these files *must not* require any change when aggregating/moving an additional level upward, i.e. any superdataset in a hierarchy references the same annex
          file of a subdataset's metadata set, any information required to fully "resolve" a subdataset's metadata in a particular superdataset, must be placed into that superdatasets's `.datalad/meta/info` (e.g. "mount point" with respect to the hierarchy). In particular, no information about the detailed state of a dataset should be in these files (commit SHAs, volatile info on remote configuration, see e.g. #1437). This means that any restructuring of subdatasets in a superdataset will only require an update of datasets upward in the hierarchy, and will never require a re-aggregation of the downward tree branches.
        - if we are concerned by the potential number of files when aggregating a huge hierarchy, we could use the same object store concept as annex (with intermediate subdirectories).
        - **the filesystem hierarchy of subdatasets is not reflected in this directory**. subdataset relationship will not be encoded in the filenames either, but must be read from `.datalad/meta/info`.
- files with aggregated metadata stay JSON-LD as before
- helpers should become available to facilitate annex metadata based tagging of dataset content based on the metadata sources supported by git annex
    - e.g. tag all "cardiac trace recordings" with a corresponding label
    - we need an API to be able to configure this and enable automatic annex metadata update when the dataset changes

## Procedures on leaf datasets (dataset with actual data)

- metadata aggregation -- converting data from an arbitrary format into our own (homogenized) JSON-LD
    - also known as "caching"
    - the results are annexed per-metadata source files in `.datalad/meta/objects` that are listed in `.datalad/meta/info`
- metadata "import" from our cached, homogenized metadata into git annex metadata -- this step must be configurable, and we need to be able re-run automatically it whenever needed (dataset changed); annex has a `annex.genmetadata` config items and also a `post-commit-annex` hook script.
    - attaching (minimal/reasonable) dataset metadata to (not into!) `.datalad/meta/info`
    - tagging individual files based on metadata
- metadata query -- load info in `.datalad/meta/objects` and `.datalad/meta/info` to build a full graph for query

## Procedures on superdatasets

- metadata aggregation -- re-annexing unaltered files from subdatasets' `.datalad/meta/objects`, transfer their availability info if possible (for this a `copy` command would be helpful). Any information about the exact state of the subdatasets must go into `.datalad/meta/info` of the superdataset
- metadata query -- identical to leaf dataset procedure

## [List of ideas and realizations from initial brain storming]

- there can be multiple native metadata source per dataset
- there is metadata for a dataset as a whole, and for individual files in a dataset
- extraction of metadata from their native formats can be costly -> cache locally
  (somewhere, in some form, not necessarily the same as for aggregation)
- aggregation of metadata in superdatasets must not require availability of an entire
  dataset hierarchy -> store intermediate aggregation levels in datasets
- metadata itself can be heavy -> distinguish between light and heavy meta data and
  use appropriate storage, e.g. put heavy aggregated metadata into annex, while keeping
  light metadata in git. Heavy vs. light is not necessarily equivalent to
  dataset-level vs. file-level
- superdataset authors must be able to filter out undesired metadata on
  aggregate, e.g. disable by metadata source
- annex provides a file/key-based key-value metadata store, but nothing for a dataset
  as a whole, or for any sub-hierarchy within a single repository. However, it could be
  coaxed into it by attaching metadata to files in a shadow hierarchy under .datalad/meta
  (presently done with JSON files already). Files would need to contain the respective
  ID of the component they are describing to yield a unique key in the annex

  Note, git-annex cannot natively handle meta data for files with content
  directly in git, and also not for submodules (mount points).

- Using the git annex metadata store could be superior to checking metadata storage files
  into git. we have tools to cut history of the git-annex branch already, while it is harder
  (but not impossible) to achieve the same for a custom solution. Generally, we don't have
  much interest in metadata history (we generate it from actual content), and the less we
  have it impact the filesystem the better
- a related issue is to expose meaningful metadata for metadata-based views in git-annex
  (i.e. all anatomical images only). This could be achieved by generating tags from native
  metadata via a dedicated parser function

Bottom line: I am debating whether to change the storage concept for metadata from custom
files with JSON-LD content to JSON-LD metadata values in annex's keystore.

**Comment by mih** [comment]
Important implication: plain Git repos have no (cached) metadata.

But I don't see a bad cost/benefit ratio when turning a plain git repo into an annex repo, at the point where metadata is desired.

**Comment by bpoldrack** [comment]
> - superdataset authors must be able to filter out undesired metadata on
>  aggregate, e.g. disable by metadata source

May be not only "by metadata source" but even key-based or sth. like that? I don't think it's important to fully implement right now, but would be nice to allow for it by design.

> But I don't see a bad cost/benefit ratio when turning a plain git repo into an annex repo, at the point where metadata is desired.

I agree.

**Comment by mih** [comment]
> May be not only "by metadata source" but even key-based or sth. like that?

By-source is actually based on a key in the current implementation. The difficulty is that metadata has an arbitrary structure and it is difficult to express filters robustly. The technicalities are much simpler.

**Comment by mih** [comment]
Important implication: metadata aggregation upwards in the hierarchy becomes impossible when there is one plain git repo in the chain.

**Comment by yarikoptic** [comment]
> But I don't see a bad cost/benefit ratio when turning a plain git repo into an annex repo, at the point where metadata is desired.

The only benefit I would see would be if there is a file to attach meta-data to, so effectively making those repos "full fledged" annex repos with some content requiring annex supported hosting.  I guess I will grasp it better as things evolve ;)

**Comment by mih** [comment]
Haha, no thanks. Nothing will evolve anywhere without a pre-approved strategy. I have prior experience with these things...

**Comment by mih** [comment]
@bpoldrack @yarikoptic Top comment has growing description of new metadata concept. Please chime in if you feel like it.

**Comment by yarikoptic** [comment]
thanks for the buzz.  I will try to get through it soon(ish).  yet thinking about leaking concern as well

**Comment by mih** [comment]
Unless I hear from you @yarikoptic @bpoldrack otherwise, I will start implementing this proposal.

**Comment by yarikoptic** [comment]
0. `I am debating whether to change the storage concept for metadata from custom
files with JSON-LD content to JSON-LD metadata values in annex's keystore. `
- at first I thought that it would lead to only cons e.g.
  - inability to publish it to non-annex-compatible portals, e.g. github
  - loosing the benefit of git compressing its objects, thus providing efficient updates to those possibly long .json files.  Every update then would require fetching entire possibly growing in size file(s)
- but in general I do not mind I guess, if we indeed start to aggregate even summaries of per-file meta-data which would lead to higher chance to leak meta-data
1.
```markdown
    - `.datalad/meta/info`: JSON file providing basic properties for all aggregated metadata files in the store (see next). This file is always present, even when no subdatasets exist.
        - **Basic dataset metadata is attached to this file using `git annex metadata`**. We might need to include some info on subdatasets in here too, in order to have useful minimal metadata for subdatasets after "`git clone`"
        - Basic dataset metadata of **subdataset** could be aggregated into the **content** of this file in a superdataset
```
- I am not entirely clear about ideas of this concept ("volatile" `.datalad/meta/info` file also with aggregated meta-data), but if I got it right I see following scalability issue:  annex will assign those aggregated fields to the 'key' of that meta-data file.  If that file to change more or less frequently, its key would be changing, meta-data would need to be duplicated/expanded for the the new key.  Might lead to explosion of the amount of storage you would need for `git-annex` branch even though many keys (one per aggregation) there then would carry copies of the aggregated meta-data.  I guess git's compression of objects would mitigate it to a good degree, but still not sure if there would be no significant impact as the history of the dataset grows.  Altogether I do not see much of positive use for adding `annex metadata` to that file, might actually get in a way (happen I decide to make a view of the annexed files based on some meta-data field which gets aggregated and then present in metadata for that file?)

- I would also prefer to have a bit more descriptive name than just `info`, e.g. if it is aggregated meta-data -- call it so, i.e. `aggregated`.

2. `.datalad/meta/objects`
 -  not sure what those `categorylabel` would be needed for, and how 'cleansing' would happen (upon reaggregation remove previous hashes?) to keep it all reasonable
 - yet to understand this hole scheme... I guess some specific example/drawing sketch could've helped

3. `helpers`
  - before all the fanciness, I am in a dire need for a basic  dataset metadata  entry helper.  ATM establishing even basic meta-data for the dataset is a dark magic.  I don't care what format it would be in, or how entered... may be it could be that we just provide a template with descriptive comments (which would be ignored when saving actual result, but better format allows for comments), so when people run smth like `datalad metadata --edit`, it would e.g. copy the template, and fire up editor, and when done (exit editor) would validate that it is all kosher, and if not -- alert to problems and provide an option to edit again, and when succeeded to be clean, aggregate/save it right away.  Would be tremendously useful.
  - assigning tags based on some filename regexps might indeed be cool/useful. We should provide clear feedback though on what files got what flags to help prevent marking files incorrectly (e.g. assigning some tag which would "facilitate" public publishing of those files, whenever they might should have not been published publicly)

**Comment by mih** [comment]
> scalability issue: annex will assign those aggregated fields to the 'key' of that meta-data file. If that file to change more or less frequently, its key would be changing, meta-data would need to be duplicated/expanded for the the new key.

Good point. So we should have a third file whose content does not change unless the datasets' own basic metadata changes. Maybe we simply put a dataset's basic metadata in this file which solves the where to put this problem. On caching/aggregation annex meta data is updated from this file. So only a change in meta data will cause a change in the key, and thereby justify an updated metadata set for the new key.

**Comment by mih** [comment]
Most of this has been implemented one way or the other in #1630. Closing.
